File: suc/swift/042.swift

import SwiftUI

// MARK: - AppTokens
// 统一的UI设计系统，方便修改比例和控件大小
struct AppTokens {
    struct Colors {
        // 将Compose的Color(0xFF...)转换为SwiftUI的Color(red:green:blue:)
        static let primary = Color(red: 0.055, green: 0.647, blue: 0.914) // 0xFF0EA5E9
        static let secondary = Color(red: 0.220, green: 0.741, blue: 0.973) // 0xFF38BDF8
        static let tertiary = Color(red: 0.490, green: 0.827, blue: 0.988) // 0xFF7DD3FC
        static let background = Color(red: 0.941, green: 0.976, blue: 1.000) // 0xFFF0F9FF
        static let surface = Color(red: 1.000, green: 1.000, blue: 1.000) // 0xFFFFFFFF
        static let surfaceVariant = Color(red: 0.878, green: 0.949, blue: 0.996) // 0xFFE0F2FE
        static let outline = Color(red: 0.729, green: 0.902, blue: 0.992) // 0xFFBAE6FD
        static let success = Color(red: 0.133, green: 0.773, blue: 0.369) // 0xFF22C55E
        static let warning = Color(red: 0.988, green: 0.804, blue: 0.082) // 0xFFFACC15
        static let error = Color(red: 0.937, green: 0.267, blue: 0.267) // 0xFFEF4444
        static let onPrimary = Color(red: 1.000, green: 1.000, blue: 1.000) // 0xFFFFFFFF
        static let onSecondary = Color(red: 0.118, green: 0.118, blue: 0.118) // 0xFF1E1E1E
        static let onTertiary = Color(red: 0.059, green: 0.091, blue: 0.165) // 0xFF0F172A
        static let onBackground = Color(red: 0.059, green: 0.091, blue: 0.165) // 0xFF0F172A
        static let onSurface = Color(red: 0.059, green: 0.091, blue: 0.165) // 0xFF0F172A
    }

    struct TypographyTokens {
        // 将Compose的TextStyle转换为SwiftUI的Font
        static let display = Font.system(size: 26, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // FontWeight.Normal 对应 .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        // 将Compose的RoundedCornerShape的dp值转换为SwiftUI的CGFloat
        static let small: CGFloat = 6.0
        static let medium: CGFloat = 10.0
        static let large: CGFloat = 16.0
    }

    struct Spacing {
        // 将Compose的Dp值转换为SwiftUI的CGFloat
        static let sm: CGFloat = 6.0
        static let md: CGFloat = 10.0
        static let lg: CGFloat = 14.0
        static let xl: CGFloat = 20.0
        static let xxl: CGFloat = 28.0
    }

    struct ShadowSpec {
        let elevation: CGFloat
        let radius: CGFloat
        let dy: CGFloat
        let opacity: Double // Kotlin的Float对应Swift的Double
    }

    struct ElevationMapping {
        // 将Compose的ElevationMapping转换为SwiftUI的ShadowSpec
        static let level1 = ShadowSpec(elevation: 2.0, radius: 4.0, dy: 2.0, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 6.0, radius: 8.0, dy: 4.0, opacity: 0.18)
    }
}

// MARK: - Data Model
// Product数据模型，遵循Identifiable协议以便在ForEach中使用
struct Product: Identifiable {
    let id: Int
    let name: String
    let price: String
    let store: String
}

// MARK: - RootScreen
// 对应Compose的RootScreen Composable函数
struct RootScreen: View {
    // 模拟Compose的remember，数据源在View初始化时创建
    let products = [
        Product(id: 1, name: "Wireless Earbuds", price: "$59.99", store: "ShopA"),
        Product(id: 2, name: "Smart Watch", price: "$129.99", store: "ShopB"),
        Product(id: 3, name: "Laptop Stand", price: "$39.99", store: "ShopC"),
        Product(id: 4, name: "Mechanical Keyboard", price: "$89.99", store: "ShopA"),
        Product(id: 5, name: "Noise Cancel Headset", price: "$149.99", store: "ShopB"),
        Product(id: 6, name: "Ergo Mouse", price: "$45.99", store: "ShopC")
    ]

    var body: some View {
        // Scaffold的替代：使用ZStack作为背景，ScrollView包裹内容
        ZStack {
            // 背景渐变色，对应Compose的Brush.verticalGradient
            LinearGradient(
                gradient: Gradient(colors: [
                    AppTokens.Colors.secondary.opacity(0.25),
                    AppTokens.Colors.background,
                    AppTokens.Colors.primary.opacity(0.25)
                ]),
                startPoint: .top,
                endPoint: .bottom
            )
            .ignoresSafeArea() // 使渐变色填充整个屏幕，包括安全区域

            // 整体内容可滚动，对应Compose的LazyColumn/LazyVerticalGrid的混合滚动行为
            ScrollView(.vertical, showsIndicators: false) {
                // 对应Compose的Column
                VStack(alignment: .leading, spacing: AppTokens.Spacing.lg) {
                    Text("Cold Gradient Price Compare")
                        .font(AppTokens.TypographyTokens.display)
                        .foregroundColor(AppTokens.Colors.onBackground)

                    // 对应Compose的LazyVerticalGrid
                    LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())],
                              spacing: AppTokens.Spacing.md) {
                        ForEach(products) { p in
                            ProductCard(product: p)
                        }
                    }
                    // 对应Compose的contentPadding = PaddingValues(bottom = AppTokens.Spacing.xxl)
                    .padding(.bottom, AppTokens.Spacing.xxl)

                    Text("Price Update Log")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.primary)

                    // 对应Compose的LazyColumn
                    LazyVStack(spacing: AppTokens.Spacing.sm) {
                        ForEach(products) { p in
                            PriceUpdateLogItem(product: p)
                        }
                    }
                    // 对应Compose的contentPadding = PaddingValues(bottom = 48.dp)
                    .padding(.bottom, 48)
                }
                // 对应Compose的Modifier.padding(AppTokens.Spacing.lg)应用于整个Column
                .padding(AppTokens.Spacing.lg)
            }
        }
        // 确保内容延伸到屏幕边缘
        .edgesIgnoringSafeArea(.all)
        // 隐藏顶部状态栏
        .statusBarHidden(true)
        // 强制使用浅色模式，对应lightColorScheme
        .preferredColorScheme(.light)
    }
}

// MARK: - ProductCard Component
// 对应Compose的Card组件
struct ProductCard: View {
    let product: Product

    var body: some View {
        // 对应Compose的Column
        VStack(spacing: AppTokens.Spacing.sm) {
            // 占位符Box，对应Compose的Box with size(100.dp)
            Rectangle()
                .fill(AppTokens.Colors.surfaceVariant)
                .frame(width: 100, height: 100)
                .cornerRadius(AppTokens.Shapes.medium)

            Text(product.name)
                .font(AppTokens.TypographyTokens.title)
                .foregroundColor(AppTokens.Colors.onSurface)

            Text(product.price)
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.primary)

            Text(product.store)
                .font(AppTokens.TypographyTokens.label)
                .foregroundColor(AppTokens.Colors.tertiary)

            // 对应Compose的Button
            Button(action: {}) {
                Text("Compare")
                    .font(AppTokens.TypographyTokens.label)
                    .foregroundColor(AppTokens.Colors.onPrimary)
                    .frame(maxWidth: .infinity) // 对应Compose的Modifier.fillMaxWidth()
                    // 对应Compose ButtonDefaults.ContentPadding.vertical = 8.dp
                    .padding(.vertical, 8.0)
            }
            .background(AppTokens.Colors.primary) // 对应Compose的containerColor
            .cornerRadius(AppTokens.Shapes.medium) // 对应Compose的shape
        }
        // 对应Compose的Modifier.padding(AppTokens.Spacing.md)
        .padding(AppTokens.Spacing.md)
        .background(AppTokens.Colors.surface) // 对应Compose的containerColor
        .cornerRadius(AppTokens.Shapes.large) // 对应Compose的shape
        // 对应Compose的elevation
        .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                radius: AppTokens.ElevationMapping.level1.radius,
                x: 0, // Compose的elevation通常没有水平偏移，或默认0
                y: AppTokens.ElevationMapping.level1.dy)
    }
}

// MARK: - PriceUpdateLogItem Component
// 对应Compose的Row组件
struct PriceUpdateLogItem: View {
    let product: Product

    var body: some View {
        // 对应Compose的Row
        HStack {
            Text("\(product.name) (\(product.store))")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface)
            Spacer() // 对应Compose的Arrangement.SpaceBetween
            Text(product.price)
                .font(AppTokens.TypographyTokens.label)
                .foregroundColor(AppTokens.Colors.secondary)
        }
        // 对应Compose的Modifier.padding(AppTokens.Spacing.md)
        .padding(AppTokens.Spacing.md)
        .background(AppTokens.Colors.surface) // 对应Compose的background
        .cornerRadius(AppTokens.Shapes.small) // 对应Compose的shape
    }
}

// MARK: - App Entry Point
// 应用程序的入口点，对应MainActivity和setContent
@main
struct PriceCompareApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}

File: suc/kotlin/042.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "056*PriceCompare*en"
private const val UI_TYPE = "ECommerce"
private const val STYLE_THEME = "Cold Gradient"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF0EA5E9)
        val secondary = Color(0xFF38BDF8)
        val tertiary = Color(0xFF7DD3FC)
        val background = Color(0xFFF0F9FF)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFE0F2FE)
        val outline = Color(0xFFBAE6FD)
        val success = Color(0xFF22C55E)
        val warning = Color(0xFFFACC15)
        val error = Color(0xFFEF4444)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1E1E1E)
        val onTertiary = Color(0xFF0F172A)
        val onBackground = Color(0xFF0F172A)
        val onSurface = Color(0xFF0F172A)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 14.dp
        val xl = 20.dp
        val xxl = 28.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class Product(val id: Int, val name: String, val price: String, val store: String)

@Composable
fun RootScreen() {
    val products = remember {
        listOf(
            Product(1, "Wireless Earbuds", "$59.99", "ShopA"),
            Product(2, "Smart Watch", "$129.99", "ShopB"),
            Product(3, "Laptop Stand", "$39.99", "ShopC"),
            Product(4, "Mechanical Keyboard", "$89.99", "ShopA"),
            Product(5, "Noise Cancel Headset", "$149.99", "ShopB"),
            Product(6, "Ergo Mouse", "$45.99", "ShopC")
        )
    }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.secondary.copy(alpha = 0.25f),
                            AppTokens.Colors.background,
                            AppTokens.Colors.primary.copy(alpha = 0.25f)
                        )
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg)
        ) {
            Text("Cold Gradient Price Compare", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.onBackground)
            LazyVerticalGrid(
                columns = GridCells.Fixed(2),
                horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md),
                verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md),
                contentPadding = PaddingValues(bottom = AppTokens.Spacing.xxl)
            ) {
                items(products) { p ->
                    Card(
                        shape = AppTokens.Shapes.large,
                        colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                        elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation)
                    ) {
                        Column(
                            modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.md),
                            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm),
                            horizontalAlignment = Alignment.CenterHorizontally
                        ) {
                            Box(
                                modifier = Modifier
                                    .size(100.dp)
                                    .background(AppTokens.Colors.surfaceVariant, AppTokens.Shapes.medium)
                            )
                            Text(p.name, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                            Text(p.price, style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.primary)
                            Text(p.store, style = AppTokens.TypographyTokens.label, color = AppTokens.Colors.tertiary)
                            Button(
                                onClick = {},
                                colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                                shape = AppTokens.Shapes.medium,
                                modifier = Modifier.fillMaxWidth()
                            ) {
                                Text("Compare", style = AppTokens.TypographyTokens.label)
                            }
                        }
                    }
                }
            }
            Text("Price Update Log", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.primary)
            LazyColumn(
                verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm),
                contentPadding = PaddingValues(bottom = 48.dp)
            ) {
                items(products) { p ->
                    Row(
                        modifier = Modifier.fillMaxWidth().background(AppTokens.Colors.surface, AppTokens.Shapes.small).padding(AppTokens.Spacing.md),
                        horizontalArrangement = Arrangement.SpaceBetween
                    ) {
                        Text("${p.name} (${p.store})", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onSurface)
                        Text(p.price, style = AppTokens.TypographyTokens.label, color = AppTokens.Colors.secondary)
                    }
                }
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFF0F9FF)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
